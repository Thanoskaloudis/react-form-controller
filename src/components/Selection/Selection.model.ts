export interface IProps {
  label: string;
  validation: () => string;
  handleCallback: (selectedId: number) => void;
  selectionData: ISelectionData[];
  fetchError: string
}

export interface ISelectionData {
  name: string
  id: number
}