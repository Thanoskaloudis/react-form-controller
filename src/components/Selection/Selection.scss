.selections {
  &--label {
    height: 5rem;
    display: flex;
    align-items: center;
    margin-left: 1.5rem;
  } 

  &--options {
    display: flex;
    flex-flow: row;
    justify-content: space-around;
    align-items: center;

    &--button {
      width: 8rem;
      height: 4rem;
      width: 8rem;
      height: 4rem;
      padding: .5rem;
      border: 1px solid #afacac;
      background-color: white;
      color: black;
      font-size: 12px;
      font-weight: 600;
      cursor: pointer;
    }

    .selected {
      border: 3px solid #7dd7d9;
      background-color: rgb(66 127 128 / 13%);
    }
  }

  &--error {
    height: 3rem;
    color: #e95454;
    display: flex;
    align-items: center;
    margin-left: 1.5rem;
  }
}