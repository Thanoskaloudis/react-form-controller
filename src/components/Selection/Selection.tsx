import React from 'react';
import { useState, useRef, useEffect } from 'react';
import '../Selection/Selection.scss';
import { IProps } from '../Selection/Selection.model';

export const Selection = (props: IProps) => {
  const [data] = useState(props.selectionData);
  const [label] = useState(props.label);
  const [optionsError, updateOptionsError] = useState('');
  const [selectedItemId, updateSelectedItemId] = useState(0);
  const [errorMessage, updateErrorMessage] = useState('');
  const [isMessageVisible, setIsMessageVisible] = useState(false);
  const timesMounted = useRef(0);

  const handleSelection = (event: React.MouseEvent<HTMLButtonElement>,id: number) => {
    event.preventDefault();

    updateSelectedItemId(id);
    props.handleCallback(id);
  };

  useEffect(() => {
    if(errorMessage.length > 0){
      setIsMessageVisible(true);
    } else {
      setIsMessageVisible(false);
    }
    
    if (timesMounted.current >= 4) {
      updateErrorMessage(props.validation);
    } else {
      timesMounted.current +=1;
    }

    updateOptionsError(props.fetchError)
  }, [errorMessage, props.validation, props.selectionData, props.fetchError]);

  let options = props.selectionData.map((option) => (
    <button
      key={option.id}
      id={option.name}
      aria-describedby="options required"
      onClick={(e) => handleSelection(e, option.id)}
      className={`selections--options--button ${selectedItemId === option.id && "selected"}`}
    >
      {option.name}
    </button>
  ));

  return (
    <div aria-labelledby="selections" className="selections">
      <label id="selections" htmlFor="selections" className="selections--label">{label}</label>
      <div className="selections--options">

        {optionsError.length > 0 && optionsError}

        {data && options}

      </div>
      {isMessageVisible ? <span id="options required" className="selections--error"> {errorMessage} </span> : ""}
    </div>
  );
};
