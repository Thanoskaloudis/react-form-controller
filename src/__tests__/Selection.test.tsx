import renderer from 'react-test-renderer';
import {render, screen, fireEvent} from '@testing-library/react'
import { Selection } from '../components/Selection/Selection';

const MockData = [
  {
    name: 'Glenna Reichert',
    id: 9,
  },
  {
    name: 'Kurtis Weissnat',
    id: 7,
  },
  {
    name: 'Mrs. Dennis Schulist',
    id: 6,
  },
  {
    name: 'Chelsey Dietrich',
    id: 5,
  },
];

describe('<Selection />', () => {
  let label = 'test';
  let validation = jest
    .fn()
    .mockImplementation(() => 'error message');
  let handleCallback = jest.fn();
  let selectionData = MockData;
  let fetchError = 'error';

  const component = renderer.create(
    <Selection
      label={label}
      validation={validation}
      handleCallback={handleCallback}
      selectionData={selectionData}
      fetchError={fetchError}
    />
  );

  it('Should render correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have the correct number of options", async () => {
    render(
      <Selection
        label={label}
        validation={validation}
        handleCallback={handleCallback}
        selectionData={selectionData}
        fetchError={fetchError}
      />
    );
    const options = screen.getAllByRole('button', {});
    expect(options).toHaveLength(4);
  });

  it('should have the correct label text', () => {
    render(
      <Selection
        label={label}
        validation={validation}
        handleCallback={handleCallback}
        selectionData={selectionData}
        fetchError={fetchError}
      />
    );
    const labelChecked = screen.getByLabelText(label, {});
    expect(labelChecked).toBeTruthy();
  })

  it('should call the callback funtion on click', () => {
    render(
      <Selection
        label={label}
        validation={validation}
        handleCallback={handleCallback}
        selectionData={selectionData}
        fetchError={fetchError}
      />
    );

    const data = MockData[0].name;
    const id = MockData[0].id;
    const option = screen.getByText(data);

    fireEvent.click(option);

    expect(handleCallback).toHaveBeenCalledWith(id);
  })
});
