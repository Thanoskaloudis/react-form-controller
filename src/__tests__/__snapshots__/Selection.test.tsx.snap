// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selection /> Should render correctly 1`] = `
<div
  aria-labelledby="selections"
  className="selections"
>
  <label
    className="selections--label"
    htmlFor="selections"
    id="selections"
  >
    test
  </label>
  <div
    className="selections--options"
  >
    <button
      aria-describedby="options required"
      className="selections--options--button false"
      id="Glenna Reichert"
      onClick={[Function]}
    >
      Glenna Reichert
    </button>
    <button
      aria-describedby="options required"
      className="selections--options--button false"
      id="Kurtis Weissnat"
      onClick={[Function]}
    >
      Kurtis Weissnat
    </button>
    <button
      aria-describedby="options required"
      className="selections--options--button false"
      id="Mrs. Dennis Schulist"
      onClick={[Function]}
    >
      Mrs. Dennis Schulist
    </button>
    <button
      aria-describedby="options required"
      className="selections--options--button false"
      id="Chelsey Dietrich"
      onClick={[Function]}
    >
      Chelsey Dietrich
    </button>
  </div>
</div>
`;
